go/host/db/batches.go

```diff
@@ -2,106 +2,121 @@ package db
 	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/obscuronet/go-obscuro/go/common"
-	"github.com/obscuronet/go-obscuro/go/common/log"
 )
 
 // DB methods relating to batches.
 
-// GetHeadBatchHeader returns the header of the node's current head batch, or (nil, false) if no such header is found.
-func (db *DB) GetHeadBatchHeader() (*common.Header, bool) {
-	headBatchHash, found := db.readHeadBatchHash()
-	if !found {
-		return nil, false
+// GetHeadBatchHeader returns the header of the node's current head batch.
+func (db *DB) GetHeadBatchHeader() (*common.Header, error) {
+	headBatchHash, err := db.readHeadBatchHash()
+	if err != nil {
+		return nil, err
 	}
 	return db.readBatchHeader(*headBatchHash)
 }
 
 // AddBatchHeader adds a batch's header to the known headers
-func (db *DB) AddBatchHeader(header *common.Header, txHashes []common.TxHash) {
+func (db *DB) AddBatchHeader(header *common.Header, txHashes []common.TxHash) error {
 	b := db.kvStore.NewBatch()
-	db.writeBatchHeader(header)
+	err := db.writeBatchHeader(header)
+	if err != nil {
+		return err
+	}
 
 	// TODO - #718 - Store the batch txs, batch hash, and batch number per transaction hash, if needed (see `AddRollupHeader`).
 
 	// TODO - #718 - Update the total transactions, once we no longer do this in `AddRollupHeader`.
 
 	// update the head if the new height is greater than the existing one
-	headBatchHeader, found := db.GetHeadBatchHeader()
-	if !found || headBatchHeader.Number.Int64() <= header.Number.Int64() {
-		db.writeHeadBatchHash(header.Hash())
+	headBatchHeader, err := db.GetHeadBatchHeader()
+	if err != nil && !errors.Is(err, errutil.ErrNotFound) {
+		return err
+	}
+	if errors.Is(err, errutil.ErrNotFound) || headBatchHeader.Number.Int64() <= header.Number.Int64() {
+		err = db.writeHeadBatchHash(header.Hash())
+		if err != nil {
+			return err
+		}
 	}
 
 	if err := b.Write(); err != nil {
-		db.logger.Crit("Could not write batch.", log.ErrKey, err)
+		return err
 	}
+
+	return nil
 }
 
 // headerKey = batchHeaderPrefix  + hash
 func batchHeaderKey(hash gethcommon.Hash) []byte {
 	return append(batchHeaderPrefix, hash.Bytes()...)
 }
 
-// Retrieves the batch header corresponding to the hash, or (nil, false) if no such header is found.
-func (db *DB) readBatchHeader(hash gethcommon.Hash) (*common.Header, bool) {
+// Retrieves the batch header corresponding to the hash.
+func (db *DB) readBatchHeader(hash gethcommon.Hash) (*common.Header, error) {
 	f, err := db.kvStore.Has(batchHeaderKey(hash))
 	if err != nil {
-		db.logger.Crit("could not retrieve batch header.", log.ErrKey, err)
+		return nil, err
 	}
 	if !f {
-		return nil, false
+		return nil, errutil.ErrNotFound
 	}
 	data, err := db.kvStore.Get(batchHeaderKey(hash))
 	if err != nil {
-		db.logger.Crit("could not retrieve batch header.", log.ErrKey, err)
+		return nil, err
 	}
 	if len(data) == 0 {
-		return nil, false
+		return nil, errutil.ErrNotFound
 	}
```

Comment:
Why 2 operations ? Has and then Get ?

Reply:
Yeah, that's the way it works apparently. Tried removing the `Has`, didn't work.

Reply:
I mean why use the two operations when `readBatchHeader` returns the value?
In other words, could the Has calling be skipped and just do the Get call ?

Reply:
But this is inside `readBatchHeader`?

Reply:
Why is the `readBatchHeader` method (yes, readBatchHeader is the method that we're discussing) doing Has (`db.kvStore.Has(batchHeaderKey(hash))`)  and then doing a Get (`db.kvStore.Get(batchHeaderKey(hash))`)? 
Can it not just do the Get ? 

Give me a shout if I'm not explaining myself properly :)

Reply:
It's not clear to me either (just copy and pasting stuff to a large extent :D ), but try removing it and you'll see. I assume Matt went through this when he originally wrote it.

Reply:
Happy for us to investigate further. But feel this is really nothing to do with my PR. It was like that before.

Reply:
It's a show-stopper for me to be honest. But I understand that it sidetracks this feature. TODO + create issue ?

Reply:
Yeah, TODO + issue is fine. I'll add.

I understand your concerns, but I think a PR review should be broadly limited to what's being changed, not whether your code changes something near something crap :)